sdk/core/scheduler/main.cc

```diff
@@ -59,6 +59,27 @@ namespace
 	 */
 	sched::Thread *futexWaitingList;
 
+	/**
+	 * Returns the boosted priority provided by waiters on a futex.
+	 *
+	 * This finds the maximum priority of all threads waiting on the futex.
+	 * Callers may be about to add a new thread to that list and so another
+	 * priority can be provided, which will be used if it is larger than any of
+	 * the priorities of the other waiters.
+	 */
+	uint8_t priority_boost_for_thread(uint16_t threadID, uint8_t priority = 0)
```

Comment:
Two problems here:

 - Stale comment from when this was taking the futex address not the thread.  That had TOCTOU issues, so now threads capture the thread ID of the thread that they're boosting when they sleep.
 - `thread->futexPriorityBoostingThread == threadID` was just `thread->futexPriorityBoostingThread`.  Ooops.

Reply:
Thinking about this, we still have one issue:

 - A acquires a PI futex.
 - B and C block on it.
 - A releases the lock and wakes *one*.
 - B wakes and acquires the PI futex.

At this point, C continues to priority boost A, not B.  I think we need to do two things to fix this:

 - When we do a futex_wake on a thread with PI waiters, we need to clear their boosted thread ID (or, rather, set it to -1) but leave them marked as doing priority boosting.
 - When we do futex_wait with a PI futex, first scan the waiter list and update all threads that are waiting on this futex to boost the new owner.

The first of these is trivial to do.  The second is tricky because it may involve recalculating the priority of multiple (possibly running) threads if we have stale boosts.  For example, if A tries to acquire a futex and boosts B, B releases the lock but *doesn't* do a wake, then C acquires the lock and then D waits on it, now A is boosting B and D is boosting C, but A should be boosting C.  I don't really want to have to collect all of the threads that might need their priorities resetting (awkward space requirement) and I don't really want to recalculate everything (awkward time requirement with interrupts disabled).

I'm not overly concerned about this because you shouldn't really use PI futexes if you don't trust the code that's doing the wait and the wakes, but it's possibly a problem.

Perhaps the right thing to do is scan the sleeping-on-a-futex list and if you find any threads waiting on this futex and priority boosting another thread (not the no-thread placeholder) then return an error from futex_wait.

Reply:
I believe the new version is now usefully correct.  We now do two update passes over the sleeping threads:

 - When waiting, we update all threads that are waiting on the current futex (with priority inheritance) to set the thread that they're boosting to the current value in the futex word.
 - When waking, we update all threads boosting that futex that are boosting the current thread to not boost anything.

There is a small window where a thread that manages to acquire a lock in between the owning thread dropping it and releasing waiters may end up not boosted, unless another thread comes in behind it.  That isn't really fixable without capturing the futex word pointer and re-checking.

Reply:
I think that small window can be closed for cooperating code by changing the "userspace" usage of PI `futex`es.... ah, and FreeBSD's `umutex` does as I expect.  Rather than merely exchanging in the `UNLOCKED` state, it CASes using an expected word containing just the calling thread ID:
https://github.com/freebsd/freebsd-src/blob/b1d3e2b771552af071bf0964233cc5a76ea25542/lib/libthr/thread/thr_umtx.h#L154

That is, since the `LockedWithWaiters` bit (in our terminology) is going to force us into `futex_wake` anyway, we may as well leave the futex locked on the way in.  That way the scheduler can promote the first waiter to owner and recompute boosts for that thread, rather than having that funny "waiters but unlocked and without waiters" state.

Reply:
I'm not sure I understand the comment-free FreeBSD code here.

It looks like there's a fast path that does a CAS to perform a locked-with-no-waiters -> unlocked transition.  If this succeeds, hurrah, we're done.  

If this fails, it then does a CAS (in a loop - presumably because this is a weak CAS?) to do a locked-with-waiters -> unlocked transition.  

This looks as if it still has the gap that another thread arriving at the mutex can see the lock in the unlocked state and acquire it in userspace without entering the sleep queue (more likely on FreeBSD, where you might have many cores trying this at once).

The other difference in the FreeBSD case is that they can rely on begin able to read the new value in the `UMTX_OP_MUTEX_WAKE2` operation to see who the new owner is if this happens.  Our `futex_timed_wake` is intended to allow asymmetric producer-consumer operations and so doesn't rely on read permission on the read side.  This means that we can't rely on being able to read the new owner value.  We could move PI wake to a different operation that does require read permission though.

Reply:
There are two things in the (comment-free, sigh) FreeBSD code:

- If the mutex is PI, then:
  - Try, exactly once, to move from locked-by-me-and-no-waiters to unlocked.
  - If that fails, jump up into the kernel.

- If the mutex is not PI, then:
  - Try, potentially many times, to move from either locked-by-me-and-{no,yes}-waiters to unlocked.
  - Look at the old lock value and jump up into the kernel if waiters was set.

The latter case is unfair exactly as in this PR: it potentially leaves the lock unowned despite there being waiters.  But that's fine, because for a non-PI mutex built around a futex, the kernel scheduler hasn't done anything.

Reply:
If I understand correctly, the FreeBSD model relies on moving the unlock into the kernel for the PI case, so you can atomically unlock-and-wake if there are waiters.  

If we're going to add something like that, I'd rather do something a bit more general.  Eventually I want to add some of the Linux bitset things so that we can move event channels out of the scheduler and make them futex operations with a compartment that allocates them and manages the set operations but lets you do fast-path waits.  That's probably also necessary for condition variables.

Alternatively we could go down the FreeBSD route and have the scheduler provide more specialised operations, but I'd prefer to keep the scheduler API surface small if we can.

Reply:
Note: The scheduler currently can take a write-only local capability to the thing that it's waking.  Any change that makes us do the unlock in the scheduler would increase the damage that a compromised scheduler could do to confidentiality (though it does need read access on the sleep path, so only slightly).

Reply:
Oh, hmm, this requires PI futexes to have a different notify interface (so that the scheduler knows that they are PI).  That's a bit awkward.

Reply:
I think I'd rather leave it and, if anyone *really* cares about closing that hole, then they can always do the transition from locked with waiters to unlocked with interrupts disabled (including the `futex_wake` call).

Reply:
That’s not the hole. The hole is that a thread that has higher or equal priority to the threat that holds the lock (while that thread is boosted) can steal the lock in the window. If it entered the queue, it would be the front one and so would acquire the lock. It won’t the able to preempt the owning thread unless it is at least equal priority to the thread that owns the lock and, with priority inheritance, the owning thread will have the same priority as the highest priority waiter.